### DISCUSSION TITLE
Multiple databases and localized data

### DISCUSSION THREAD

**Discussion by majkinetor** [1Remote/1Remote#414] [Ideas]
Once remote database is not accessible you can no longer use its connections. We should have a local copy of all remote databases. Connection availability doesn't necessarily have anything to do with database availability - for example, database may be down for administrative reasons, and people could still work with the last cached copy.

### Multiple databases

Create new local database for each remote database. This could be done in current local sqlite (`1Remote.db`) or in separated files (`<Database>.db` from Database settings. Separated sqlite make more sense since then your team member can send others his own local cache and connections will just work. This can happen if, for example, remote database is currently not available and you never connected to it before. With single sqlite database, he can't do that without sending others his private connection data too.

Having multiple databases provides more options even in no remote database scenario - you can for example separate connections in multiple local databases for whatever reason - keeping them grouped per project, customer (if you are consultant) etc. Implementing multiple databases for remote database will automatically enable this option too, with minimal additional development.

With multiple local databases, 1RM should keep metadata about last successful connection in its own database. Since all databases are cached in memory too, we now have potentially 3 places where database is: memory, local file system, remote. Only memory is always accessible since even local database can get locked by external process.

Local FS and remote can be marked as *authority* (memory can never be authority) which determines how 1RM deals with disconnection. So original 1Remote.db is authority and MySql remote db is authority but not is local copy. 1RM keeps in memory metadata about last successful connection and next connection-retry time to authority (for both local or remote databases). When it succeeds, it should also store that in the local FS database (be it authority or not) and disable connection-retry timer.

Once authority is not available, editing connections coming from it should be disabled (they become read-only, for example by disabling *Save* button in connection editor.

### Localized data for remote connection

Based on discussion #392 there should be need for local alternation of remote connections. Original idea was for allowing custom username - team members will either have their own domain names or local OS names. Others mentioned that other connection attributes could be localized too, such as *Script before connect* which is a good point.

Given that one may create a local copy of remote connection, I think this should be optimized for the main use case - custom credentials. Default credential is IMO not flexible enough in most scenarios particularly in common case where team members get personalized credentials to remote resource. This means that we should probably allow for both - Create default credential in 1RM configuration (e.g. domain username) and allow associating remote connection with local credential. This could be done on first run when specifying empty username on remote connection. 1RM could for example add another connection field in the local cache.

### Multiple databases

1. Add new option in global configuration Databases - `SQLite`
    1. By default system comes with one default SQLite named `1Remote` but user can add another one
1. For each remote database create local database
    1. Just as with local, allow path to be specified for local cache, by default it should use local directory and and have the same name as Database
1. By default, keep databases in `db` directory in order not to pollute root file system (no matter is it portable or roaming) and since we can have many databases now. This is the most natural place for it others will rarely touch.
    1. Add additional sugary option `Open database directory`.
1. When connecting to remote database, add new local copy using SQLite.
1. On the startup iterate all available *local* databases (be it authority or not) and add them to the connection list using groups
1. Add UX options:    
    1. Collapse/uncollapse all 
    1. Number of connections next to the database name in the list (e.g. `1Remote [35, 3 selected]`)

### Forks

1. Add `Local Json` in `Servers` table in local copy of remote database
1. Provide `Save Local` button apart from `Save`, and keep any edited *differences* there (1RM already has connection diff due to the bulk edit attribute comparison)
1. Local configuration is always used over remote one, if it is defined
1. Add an option option `Clear Local` that empties entire `Local Json`
    1. It would be good (but not mandatory) for one to have *point of view* UX - *remote* vs *remote with local* for any attribute that has local override so he can switch to see a difference. This feature could have localized reset (clear only that attribute from the Local Json, not everything).

This is basically a fork, and the main benefit is that connection attributes that are not locally touched can still be synced from remote and affect local database.

### Domain credential

Regarding credentials, there is also a merit for connection maintainer to specify AD/LDAP option remotely as this is widely used - I would do that for number of my servers, specify that connection uses teams domain name. If we have such an option ( e.g. *Use local domain credentials*) one would still need to provide in 1RM configuration his local domain credentials as I think there is no way to find out his password other then [mimikatz](https://adsecurity.org/?page_id=1821).

1. Add in configuration editor a new option `Use local domain credentials`
1. Provide new global configuration a new section `Global credentials`
    1. Make one predefined value `Local domain credentials` that user should setup
1. If remote connection has option in 1. checked, use option in 2.1 if set, otherwise, ask for credentials on first connection
    1. User can choose any credential from global store or input a new one
    1. Remember credentials in `Local Json` if they are input by the user
        1. If global store credential is used, remember it by `Name` but not copy username/password in order for that user to be able to change them globally afterwards

---

**Comment by VShawn** [comment]
> Create new local database for each remote database.

Creating a new local database for each remote database can compromise security, as it may allow users with read-only permissions (who cannot view the connection password stored in MySQL) to access passwords if data is cached in the local database.


> send others his own local cache ... This can happen if, for example, remote database is currently not available 

Sending cache between users sounds very strange.

And most of the data stored in MySQL is related to the team's connections, which are likely to be internal IP addresses. When a database connection is lost, it is possible that a user has left the Wi-Fi range of the company. In this case, it is reasonable to hide all connections after the database connection is lost.

**Comment by majkinetor** [comment]
> Creating a new local database for each remote database can compromise security, as it may allow users with read-only permissions (who cannot view the connection password stored in MySQL) to access passwords if data is cached in the local database.

What do you mean? You already have access to remote database. I didn't think _exact_ copy, but the copy of the stuff you already can access and view. If you can't access a password, you wont be able to access it in your local cache too.

> Sending cache between users sounds very strange.

I don't think so. It might be cache, but its still valid database. There is no security issue in either way because user that wants to do that, can do that anyway, as he has access. Contrary, if remote db is down, and you know that your team member has accesss to it or will get it anyway, you can just send him the database and enable him to work on connections.

> And most of the data stored in MySQL is related to the team's connections, which are likely to be internal IP addresses. 

Thats assumption. Users will do that for sure. Also, server or database itself may be down (and will be down during maintenance), not the VPN to the company infrastructure. 

> When a database connection is lost, it is possible that a user has left the Wi-Fi range of the company. In this case, it is reasonable to hide all connections after the database connection is lost.

Like I said, user that wants to share company secrets will do that while he is online. You can't stop that. You can make his decision just a little bit easier or little bit harder, but neither thing will stop him if there is a will. However, IMO, we miss bigger oportunity here, for really cool features that many will certainly use:

1. If you want to use connection when database admin fucked up everything, you can do it. It can even serve as backup.
1. If you want to fork remote connection, this allows it, and you still follow the remote changes of locally unchanged stuff or even revert (like typical git branch) 
1. And since you already have multiple local databases, why not adopt it all the way, even for multiple local files (it changes nothing in implementation, its just more stuff for the same amount of work)

If you are still concerned about "security", you can make some additional options, like remote db setting that 1RM will look into to decide if databas can be cached locally or not. You already expressed concerns about passwords before (#317) and this is not different. _"It is reasonable to hide all connections after the database connection is lost"_ is not security, its wishfull thinking. I will grab things from RAM memory for sure, or will just go each connection and print screen it (even without 1 minute of automation).

**Comment by VShawn** [comment]
I once again believe that if the reason is "database may be down," caching online mysql data to a local sqlite is pointless:

1. Firstly, it's the responsibility of the DBA to ensure the stability of the online database, rather than relying on 1RM to consider how to restore data if it get messed up. backup of databases is something every DBA should do.
2. Since the online database is stable, caching it to a local file to make ues of the data becomes meaningless. The only situation where the database goes offline is when the user disconnects from the network, and it's natural not to be able to use the resources on the network.
3. There is a potential data consistency issue: in extreme cases, an admin who has been offline for a long time could overwrite the latest online data with old cached data due to a mistake editing. Unless we build a tool like Git for branch management, which I think it's overkill.
4. And security issues: currently users with read-only permission can only load server passwords into memory (encrypted strings), they can use it for connect but can not see the plaintext. And at least it's difficult to read psw from memory. but once they are cached to the local database, even a computer novice can copy the password from local database out (assuming your goal is to allow users to use connections when db offline, then you must cache passwords).

Finally, suppose a scenario where I bring my own laptop to work during the day and continue personal development on it at home in the evening. If I use your solution, then my computer would cache the company's data locally, but I definitely don't want to face more than 30 connections from the company after I get home. Therefore, after the online database goes offline, its projects should be hidden.

Of course, the local cache you proposed has other meanings, such as allowing each client to customize using their account password to connect to a server based on the data in the online database. The issue may be on how to handle conflicts between the online database and local data, such as when a connection in MySQL has already provided an account & password; how to add personal account passwords locally for it and how to modify the account password in MySQL after adding a local account password conflict to the online one, and how to switch between them. I think these issues can only be resolved through well-thought-out UI design which I have no good design yet. We may need some flowcharts and prototype diagrams for discussion to avoid any misunderstandings of our views.

**Comment by majkinetor** [comment]
> Firstly, it's the responsibility of the DBA to ensure the stability of the online database, rather than relying on 1RM to consider how to restore data if it get messed up. backup of databases is something every DBA should do.

What does that even mean?  You want to educate the world on how to be succesifull DBA? :) Its respoinsiblity of the application to be **resilient** and to work, if it can work.

> there is a potential data consistency issue

Easily solvable by looking into last change date of the database, and if it is in the past compared to the cache, you ditch the cache and recreate it (after all, its just a cache). This will make forks lost but nothing is perfect and this shouldn't be frequent thing, if at all (but easily fixable even in that case, but I wouldn't go there).

> Since the online database is stable

Database may be down for maintaince on some places regularly (like database updates, security patches etc.). But that is irrelevant. I don't assume the database will be constantly down, but VPN connection is definitelly consntatnly down. With reconnect we have now this is less of a problem, but still, you can't use database while VPN is down, and that is artificial limitation. Scenario is VERY easy to come up with - what if I want to keep cloud connections in my local database for example. Amazon/Google almost always work while your on premise database or VPN may not.

> And security issues: currently users with read-only permission can only load server passwords into memory (encrypted strings), they can use it for connect but can not see the plaintext. ... And at least it's difficult to read psw from memory.

Its not that difficult, this is security via obscurity which doesn't solve security issue.  

>If I use your solution, then my computer would cache the company's data locally, but I definitely don't want to face more than 30 connections from the company after I get home. 

Like I said, you can forbid cache on remote side if that is your concern, and all problems solved. Its enough to put something in config schema like `cache: not allowed` and we are done. You can do that actually from both side: both user and remote db can decide on their own. Why enforce that to users that have better things to do or do not care or have different security protections. **The main point is, you don't want to decide on users but give them options and let them decide**. In your 30 connections case, I can simply choose not to cache that db, I can collapse it so not to view it, or I can delete local cache if I accicdently used it, and all take 1 sec to do.

> Of course, the local cache you proposed has other meanings, such as allowing each client to customize using their account password to connect to a server based on the data in the online database. 

Not only a password, it can be done without it, but everything. You proposed changing everything may have a value, I am just thinking on how to do that :)

> The issue may be on how to handle conflicts between the online database and local data, such as when a connection in MySQL has already provided an account & password; 

Like I stated in original post, there is no conflict - if you have local fork, it has priority, even if remote data changes. I proposed we can have visibility on changes by having 1RM show exclamation marks when it detects it, but that is really nitpicking, as how hard can it be to just reset your local changes and do them again in single connection. But easy to do anyway, and you already implemented connection diff for bulk changes.

So let me bring the benefits again, that may be lost in discussion:

1. Having multiple local databases is IMO great feature - perhaps you are consultant and want to keep different company stuff separated - perhaps you are organisational freak, perhaps some stuff are more important to you then others and have different secruity standards (lets say you want to keep it on encrypted disk that might not be always available), who cares. Those features can get improved a lot too if we ever want that (config options for various stuff, like don't allow bulk updates across databases, show only current database tags when autocompleting tags on connection etc.)
2. If we support multiple local database, we are 1 afternoon away from having local cache of remote db and with it features like constant conectivity, easy backup and "forks"
3. The system is consistent and easy to understand. I think its also probably not that big effort to implement, but you tell me. I would give it a week or so. Everything else is UX we can find solution for.


I am not pushing for this solution too, its currently just a suggestion, but I think I like it 😃

**Comment by VShawn** [comment]
> You want to educate the world on how to be succesifull DBA? :)

I not intend to do this, but it do the responsibility of the DBA to ensure the the database wokrs.



>> And security issues: currently users with read-only permission can only load server passwords into memory (encrypted strings), they can use it for connect but can not see the plaintext. ... And at least it's difficult to read psw from memory.

> Its not that difficult, this is security via obscurity which doesn't solve security issue.

Yes it did not solve security but local database make password security worse.

> if you have local fork, it has priority

What you mean? 

Like: yesterday, I got a local fork and made private customizations. Today, due to server IP migration, the HOST field on MySQL was modified by the admin. What will happen to the local fork at this point?

1. If the fork is overwritten by the data on MySQL, then the customization project from yesterday will be lost. I believe that data loss is not acceptable.
2. If the fork continues to remain unchanged, with an outdated HOST in it, and the local fork has a higher priority, I will never be able to connect to this connection.

Unless we record the update time of each field, any of the above paths will result in issues.

---

What I want to express is that this plan can be pushed forward. However, there are still many details that need to be considered, especially since I don't like resolving conflicts between local forks and remote repositories. Instead, I prefer to only store customized content locally (such as personal account passwords), and enable the "connect" feature to override remote configurations with the customized configurations stored locally.

Additionally, it may be my misunderstanding, but in my view, the local cache you are currently designing doesn't seem to be significantly different from synchronizing a shared SQLite through sync for a team: both of them can keep work when the online service is down, and we can also make side configurations for the customize.

What I expect fot is:

- After 1RM is launched, it should connect to MySQL successfully at least once before being able to use the connections in MySQL. It should be able to use after disconnecting (this has already done in #395).
- The local cache only store customized content rather than all forks.

---

## BTW

> If we support multiple local database,

We did have multiple local database supported, I just hide the option

![image](https://github.com/1Remote/1Remote/assets/10143738/7cd78869-30e3-4f83-ad1f-fd2611aa2a39)


> like don't allow bulk updates across databases

Well it allow bulk edit across databases now :)

**Comment by majkinetor** [comment]
> Like: yesterday, I got a local fork and made private customizations. Today, due to server IP migration, the HOST field on MySQL was modified by the admin. What will happen to the local fork at this point?

Your local fork most definitelly didn't include the HOST field changes. It is not good idea to change those fields, why would you (although it would be possible). So lets say you chagend credentials, color, tag and before script (sounds tootally legit use case). Now if remote side has another color, AT ANY TIME, your own color takes presidence. The same with other attributes. You have to clear the field (or all fields) to get remote version again. Sounds OK, IMO. Especially if 1RM marks remote changes like this:

![image](https://github.com/1Remote/1Remote/assets/85767/47fee710-8ffe-4e9f-a9c4-e16e68d93771)

This would mean there are differences, by hovering a mouse above 1RM could offer option to reset this field (and also have 1 reset for all next to Save button). I beleive that this is easy to implement as you already did implement connection diff for bulk editing. Now, the problem is how to treat arrays, tags for example. Do you merge them with remote? But, I wouldn't do that. If you added your thing, the thing is locked for you only, even for arrays, just reset it and add your own stuff again if you want it, not a big deal. 

> Unless we record the update time of each field, any of the above paths will result in issues.

Not necessary, just 1 last change for entire connection (which we should have anyway). Then if remote last change date is newer then local (not counting your fork changes) you can diff the entire thing and show the differences. 

Or it could even be less extensive, just show ! if ANY change is detected, but allow user to see remote connection somehow without local forks, then he can copy paste changed attributes to local fork for example. Different UX, but how much this feature will be used need to be measured. I wouldn't overcomplicate it until wee see usage in production (reminds me, we need to have telemetry system for that and some infrastructure, similar to MS store thing, but that is another story we will talk about in the future).

> Additionally, it may be my misunderstanding, but in my view, the local cache you are currently designing doesn't seem to be significantly different from synchronizing a shared SQLite through sync for a team

You need to setup sync. This is out of the box without hessle. 

> The local cache only store customized content rather than all forks.

What do you mean _all the forks_? Customized content is the fork, by definition. You mean all the remote connections? You would keep in local cache only what I refered to 'Local JSON'  (when I say fork, I refer to local changes on single connection, so forks are collection of changes, 1 for each changed remote connection=

So, how are you keeping remote db cache now, since it exists after first connection ?

**Comment by majkinetor** [comment]
Well now to sumarize the story so far :)

1. We actually have multiple local databases, lol, so unhide them and present them as groups the same as with remote
2. We actually have local cache, just not mapped to local database, but somewhere...

Good story 🏅

**Comment by VShawn** [comment]
yes, We actually have local cache in RAM, cache will be disposed after 1RM exit. 

And 1RM have local cahced files:

![image](https://github.com/1Remote/1Remote/assets/10143738/81a528f4-3e8d-4998-af70-3608be607a52)

ConnectionRecords.json keep the last coconnect time of each connections, in order to make launcher order by connect time.

Locality.json keep `group order` and `group is expanded or not when app start`

**Comment by majkinetor** [comment]
So its all like I already said, just without a plan and extra options :)

OK, give it a more thought and if we do not do this, lets just fix the local credentials problem and forget about forking other fields.

**Comment by majkinetor** [comment]
Most notably, having it in RAM will make it dissapear upon crash, restart, upgrade.

**Reply by VShawn** [reply]
Yes that what I want it to be.

**Comment by VShawn** [comment]
1. As a company leader, do you want your employees to carry laptops with internal server cache to Starbucks to drink coffee? The effective scope of team data should be within the team itself, rather than having the cache of this data accessible outside the company. This is not even driven by security concerns, but rather the compliance requirements of work processes.
2. You mentioned that local cache can serve as a backup for online data. The why do we still need MySQL? We could just use SQLite and synchronize it with cloud storage for each employee. And cloud can automatically save every version of the data.
3. If you disconnect from the online database, can I assume that your network has failed? In that case, even if you have cached the data from MySQL, you won't be able to access the corresponding network, making the cache practically meaningless.

**Reply by majkinetor** [reply]
1. We already discussed that this can be optional - you could mark remote db as non-cachable if this is concern. Honestly, as a company leader, I understand that I can't prevent people of using 1RM while in Starbucks local cache or not, so security is on them anyway, you are making it somewhat harder for malicious actor, but its far from show stopper. In short, its good to have options when I don't care about Starbucks as I have other security measures in check, like MAC address or location check.
2. That is a good question - the only reason we need mysql is because we can't do concurent changes on file system without reimplementing database :) We don't need mysql though - equivalent would be to use sqlite with server process in front of it that can accept sql, liike [rqlite](https://rqlite.io). But its even questionable how often one changes remote connections and you could even go with file system synchronization, although we would need to test that (its valid question only for hard core enterprise usage with thousands of servers, but I am pretty sure 1RM would collapse long before that)
3. I am repeating myself but remote database may not work too for X reasons. For example you can keep your cloud servers in remote database and if remote database goes down, you can still access remote servers. I imagine this is even typical usecase for majority of startups which do use cloud servers but have limited man power for database maintenance. I am sure users would think of bunch of more use cases.

**Reply by VShawn** [reply]
> you are making it somewhat harder for malicious actor

Yes, raising the barrier for malicious actions is sufficient. Just as we know that current door locks can be easily picked by thieves, we still buy a door for our homes to reduce the probability of burglary.

In fact, in 1Remote team usage, only a few people should have the ability to operate the MySQL database. Most team members should only have read-only permissions, and they even have no viewing privileges. However, the idea of local cache can bypass these viewing restrictions, allowing potential thieves to use any SQLite viewing tool to see server information, just like a thief walking into your home without a door.

> because we can't do concurrent changes on the file system.

MySQL is better than sqlite in this part, but noted 1Remote's MySQL also has issues with phantom reads and writes, and I don't want to solve them right now.

---

let's talk about the benefits of providing local caching for MySQL:

1. Continued usage when the MySQL service is offline.

	my plan is that once disconnected from MySQL, users should not be able to make use of the server information stored in it. ANd I am curious why you always assume that MySQL goes offline for some reason, yet the servers within it continue to function normally? 

	AND Yes, indeed, some companies may not have the resources to maintain their MySQL databases. But would such companies have dedicated personnel to maintain 1Remote's MySQL data? Would there be someone organizing the deployment of 1Remote? In such companies, employees can simply use the personal version of 1Remote, that's enough.

	AND from my perspective, even the MySQL deployed in my Synology NAS can work 24/7 (I have never maintained it). If a company's maintenance level is worse than my Synology NAS, then they probably don't need a remote server management tool either.

	AND The idea of not repairing the MySQL service after a failure and instead continuing to work by sending local cache files to 1Remote is itself very strange IMO.

2. Fork remote connection.

	 Providing an overwrite function is sufficient, there is no need to cache data from MySQL, only the user's newly added data needs to be stored.

**Reply by majkinetor** [reply]
> Yes, raising the barrier for malicious actions is sufficient.

This is called security over obscurity basically and is [innefective](https://www.okta.com/identity-101/security-through-obscurity/)

>  Just as we know that current door locks can be easily picked by thieves, we still buy a door for our homes to reduce the probability of burglary.

You don't reduce the probability of burglary with dors but for ad hoc harresement, if you have valuable stuff and you count on dors to protect you, you don't know what you are doing.

And like I said, all problems solved if cacheing its optional. 

> MySQL is better than sqlite in this part

Not for our usecase, sqlite is way better IMO as its quicker to setup (rql is one executable for example) and trivial to maintain and backup but on the flip side you lose integrated db security. Using parallel writes to sqlite is OK for this usecase (unless you have hundreeds of writes per second which we will never have)

> But would such companies have dedicated personnel to maintain 1Remote's MySQL data?

Yes, because it doesn't require dedicated personnel team, it can be 100% collaborative. Also, you really think that maintaining 1RM key value list is the same as maintaining mysql server? Its also not about maintenance only, the database can be in private space and resources it holds can be in public space with different dynamics of accessibility.

> Would there be someone organizing the deployment of 1Remote? 

Nobody needs to organize deployment of 1RM - you install it yourself via any way possible, then you add remote database yourself from the company wiki or something.

> let's talk about the benefits of providing local caching for MySQL:

1. You forgot something we just discovered: slow/glitch startup. It will just get way worse with more then 1 remote database. 
2. There is also a local backup with mysql server that you don't have now - if your mysql is destroyed, your connections are destroyed. Do you have a backup of your synology's mysql? If so, do you regularly test it? :) 
3. There is also potential for database sharing by users (which you think is a problem, but I think is a feature if done with some protections like optinal caching)

Anyway, I think this is enough to be said about this idea. I can just add to enumeration here in the future when we encounter new cases that would work easier/better if we had this feature. Its not that I am hard core beleiver in it, especially since you are not :)

**Comment by majkinetor** [comment]
@VShawn it just happend to me as I suspected:

<details><summary>Screen</summary> 

![image](https://github.com/user-attachments/assets/0e9162a0-3dac-4dd6-9275-911d339dbf01)
</details>


Everything in the company works except 1remote DB, so now I am blind and dysfunctional on the last day of the year (the most important financially). I don't have access to mysql server to fix this and people that do are on vacation. 

If I had a local cached copy, I would be able to work normally. 

My company is extremely agile when something doesn't work, so if it happened here (first time now after a year), it will happen all the time with other companies.

**Reply by VShawn** [reply]
Is it 1Remote having issues or is it your MySQL acting up?

If it's your MySQL, then I still believe that 1Remote doesn't need local cache. Server that requires local caching should be stored in a local SQLite database.

**Reply by majkinetor** [reply]
There was a problem with MySQL connectivity.

> Server that requires local caching should be stored in a local SQLite database.

I need all of them, hundreds. You are suggesting here that this feature should be, instead, done by hand when needed, which only means that there is no associated security with current choice (which was about security)
